lib/templates.php

```diff
@@ -0,0 +1,74 @@
+<?php
+/**
+ * Register Gutenberg core block editor templates.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Registers Gutenberg core block editor templates.
+ */
+function gutenberg_register_templates() {
+	if (
+		get_option( 'gutenberg-experiments' ) &&
+		! array_key_exists( 'gutenberg-full-site-editing', get_option( 'gutenberg-experiments' ) )
+	) {
+		return;
+	}
+
+	register_post_type(
+		'wp_template',
+		array(
+			'labels'       => array(
+				'name' => __( 'Templates', 'gutenberg' ),
+			),
+			'show_in_rest' => true,
+		)
+	);
+
+	$template_query = new WP_Query(
+		array(
+			'post_type' => 'wp_template',
+			'name'      => 'single-post',
+		)
+	);
+	if ( ! $template_query->have_posts() ) {
+		wp_insert_post(
+			array(
+				'post_type'    => 'wp_template',
+				'post_name'    => 'single-post',
```

Comment:
Should the name perhaps map directly to the template hierarchy? So instead of `single.php` should this be just `single`?

Reply:
There are `single-post.php` and `single-$posttype.php` templates in the hierarchy too.

Here's a stab at how to replace them dynamically: #4659

Reply:
Oh right... `single-post` is `single-$posttype` 🙂

Reply:
Yes, that PR is what I had in mind.

Reply:
Yeah, although I think `single-post` is treated as a separate special case in the deeper parts of WP.